### PR TITLE
Chebychev factory

### PR DESCRIPTION
**Pull request by regislebrun**
1) Added the ChebychevFactory class.

2) Improved the IteratedQuadrature class.
+ Now, it calls the GaussKronrod class directly in the 1D case.

3) Improved AdaptiveStieltjesAlgorithm
+ Now it takes explicitely into account the singularities in the PDF of the measure.

4) Added PDF singularities accessor to 1D distributions.

5) Added the HistogramPolynomialFactory class.
+ It implements a very efficient and very accurate dot-product to compute orthonormal polynomials for Histogram distributions

6) Improved the parameter management in Histogram

7) Fixed a bug in JacobyFactory
+ There was a sign error on a coefficient when \alpha\neq \beta.

### CODE REVIEW

File: python/test/t_ChebychevFactory_std.expout [hunk 1]
```diff
@@ -0,0 +1,13 @@
+hermite= class=HermiteFactory measure=class=Normal name=Normal dimension=1 mean=class=NumericalPoint name=Unnamed dimension=1 values=[0] sigma=class=NumericalPoint name=Unnamed dimension=1 values=[1] correlationMatrix=class=CorrelationMatrix dimension=1 implementation=class=MatrixImplementation name=Unnamed rows=1 columns=1 values=[1]
```
Comment (sofianehaddad) [inline python/test/t_ChebychevFactory_std.expout]: hmmm hermite?? something is going wrong here...

**Comment by jschueller** [comment]
no change in StandardDistributionPolynomialFactory ?

File: python/src/HistogramPolynomialFactory_doc.i.in [hunk 2]
```diff
@@ -0,0 +1,55 @@
+%feature("docstring") OT::HistogramPolynomialFactory
+"HistogramPolynomial specific orthonormal univariate polynomial family.
```
Comment (jschueller) [inline python/src/HistogramPolynomialFactory_doc.i.in]: Histogram specific

File: python/src/HistogramPolynomialFactory_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    If `parameters_set == ot.JacobiFactory.PROBABILITY`: default shape
+    parameter :math:`k > 0` of the :class:`~openturns.Gamma` distribution.
+
+    If `parameters_set == ot.JacobiFactory.ANALYSIS`: alternative shape
+    parameter :math:`k_a = k - 1 > -1` of the :class:`~openturns.Gamma`
+    distribution.
+parameters_set : int, optional
+    Integer telling which parameters set is used for defining the distribution
+    (amongst `ot.LaguerreFactory.ANALYSIS, ot.LaguerreFactory.PROBABILITY`).
+
```
Comment (jschueller) [inline python/src/HistogramPolynomialFactory_doc.i.in]: parameters are first, width, height

File: lib/src/Uncertainty/Distribution/Histogram.cxx [hunk 4]
```diff
@@ -451,6 +451,21 @@ Histogram::HistogramPairCollection Histogram::getPairCollection() const
   return collection;
 }
 
+/* Get the PDF singularities inside of the range - 1D only */
+NumericalPoint Histogram::getSingularities() const
+{
+  const UnsignedInteger size = width_.getSize();
+  // Here we know that size > 0
+  NumericalScalar x(first_ + width_[0]);
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Histogram.cxx]: NumericalScalar x = first_ + width_[0];

File: lib/src/Uncertainty/Distribution/TruncatedDistribution.cxx [hunk 5]
```diff
@@ -525,6 +525,22 @@ NumericalSample TruncatedDistribution::getSupport(const Interval & interval) con
...
+/* Get the PDF singularities inside of the range - 1D only */
+NumericalPoint TruncatedDistribution::getSingularities() const
+{
+  NumericalPoint singularities(0);
+  NumericalPoint nontruncatedSingularities(distribution_.getSingularities());
+  const UnsignedInteger size = nontruncatedSingularities.getSize();
+  if (size == 0) return singularities;
+  for (UnsignedInteger i = 0; i < size; ++i)
+    {
+      const NumericalScalar x(nontruncatedSingularities[i]);
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/TruncatedDistribution.cxx]: use operator=

File: lib/src/Uncertainty/Algorithm/OrthogonalBasis/HistogramPolynomialFactory.cxx [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  NumericalPoint weights;
+  const NumericalPoint nodes(legendre_.getNodesAndWeights(n + 1, weights));
+  const NumericalScalar first = p_histogram_->getFirst();
+  const NumericalPoint width(p_histogram_->getWidth());
+  const NumericalPoint height(p_histogram_->getHeight());
+  NumericalScalar a = first;
+  NumericalPoint dotProduct(2);
+  for (UnsignedInteger i = 0; i < width.getSize(); ++i)
+    {
+      const NumericalScalar ab = width[i];
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/OrthogonalBasis/HistogramPolynomialFactory.cxx]: brackets are not indented throughout the rest of the code

**Comment by jschueller** [comment]
nice! 7f56da8 will be backported to the 1.8 branch

File: ChangeLog [hunk 7]
```diff
@@ -7,6 +7,7 @@
 ==== Major changes ====
 
 ==== New classes ====
+ * HistogramPolynomialFactory
```
Comment (jschueller) [inline ChangeLog]: ChebychevFactory can be mentioned too

File: lib/src/Uncertainty/Distribution/TruncatedDistribution.cxx [hunk 8]
```diff
@@ -525,6 +525,22 @@ NumericalSample TruncatedDistribution::getSupport(const Interval & interval) con
   return distribution_.getSupport(getRange().intersect(interval));
 }
 
+/* Get the PDF singularities inside of the range - 1D only */
+NumericalPoint TruncatedDistribution::getSingularities() const
+{
+  NumericalPoint singularities(0);
+  NumericalPoint nontruncatedSingularities(distribution_.getSingularities());
+  const UnsignedInteger size = nontruncatedSingularities.getSize();
+  if (size == 0) return singularities;
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/TruncatedDistribution.cxx]: this test is redundant with the one in the loop : the body is not executed and singularities is returned anyway.

**Comment by sofianehaddad** [comment]
f72d892 is also a bugfix that could be backported to the 1.8 branch

**Comment by jschueller** [comment]
there's no unit-test for HistogramPolynomialFactory, although it's tested through StandardPolynomialFactory, but as you added one for ChebychevFactory so...

**Comment by jschueller** [comment]
+1

**Comment by sofianehaddad** [comment]
+1

**Pull request merged**